File: trilogy/io/contract.py
"""What trilogy may ask a source script for, and how that ask is honored.

A request is satisfied one of two ways. If the source function declares a
matching parameter it is bound and the function owns it -- an API-backed source
can push the filter to the API, which is the point of having a contract at all.
Whatever the function does not declare, ``apply`` enforces on the Arrow stream.

So the contract is always honored; declaring a parameter only changes *where*.
Adding a field here means adding a fallback below, after which existing scripts
keep working and silently gain the feature.
"""

from __future__ import annotations

import inspect
import json
import re
from collections.abc import Callable, Iterator, Mapping
from dataclasses import dataclass, field, fields, replace
from typing import Any

import pyarrow as pa

from trilogy.io.errors import ContractError

CONTRACT_VERSION = 1

# Bumping this is a wire-format change; the rust crate parses the same strings.
OPERATORS = ("!=", ">=", "<=", "=", ">", "<", "not in", "in", "like")

_FILTER = re.compile(
    r"^\s*(?P<column>[A-Za-z_][A-Za-z0-9_.]*)\s*"
    r"(?P<op>!=|>=|<=|=|>|<|\bnot\s+in\b|\bin\b|\blike\b)\s*"
    r"(?P<value>.*?)\s*$",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class Filter:
    column: str
    op: str
    value: Any

    @classmethod
    def parse(cls, text: str) -> Filter:
        match = _FILTER.match(text)
        if not match:
            raise ContractError(
                f"Could not parse filter {text!r}. Expected '<column> <op> <value>' "
                f"where op is one of {', '.join(OPERATORS)}."
            )
        op = " ".join(match.group("op").lower().split())
        return cls(match.group("column"), op, _parse_value(match.group("value")))

    def render(self) -> str:
        return f"{self.column} {self.op} {json.dumps(self.value, default=str)}"


def _parse_value(raw: str) -> Any:
    """JSON first, so numbers, lists and null work; bare words stay strings."""
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw.strip("'\"")


@dataclass(frozen=True)
class Sort:
    column: str
    descending: bool = False

    @classmethod
    def parse(cls, text: str) -> Sort:
        column, _, direction = text.partition(":")
        normalized = direction.strip().lower() or "asc"
        if normalized not in ("asc", "desc"):
            raise ContractError(
                f"Could not parse sort {text!r}. Expected '<column>' or "
                "'<column>:asc' / '<column>:desc'."
            )
        return cls(column.strip(), normalized == "desc")

    def render(self) -> str:
        return f"{self.column}:{'desc' if self.descending else 'asc'}"


@dataclass(frozen=True)
class SourceRequest:
    limit: int | None = None
    columns: tuple[str, ...] | None = None
    filters: tuple[Filter, ...] = ()
    order_by: tuple[Sort, ...] = ()
    since: Any | None = None
    partition: Mapping[str, str] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "limit": self.limit,
            "columns": list(self.columns) if self.columns else None,
            "filters": [f.render() for f in self.filters],
            "order_by": [s.render() for s in self.order_by],
            "since": self.since,
            "partition": dict(self.partition),
        }


#: Parameter names a source function may declare to take ownership of a field.
PUSHDOWN_PARAMETERS = tuple(f.name for f in fields(SourceRequest))


def pushdown_parameters(fn: Callable) -> tuple[str, ...]:
    """Which contract fields ``fn`` declares -- the whole request if it asks."""
    try:
        signature = inspect.signature(fn)
    except (TypeError, ValueError):
        return ()
    if _request_parameter(signature) is not None:
        return PUSHDOWN_PARAMETERS
    return tuple(n for n in PUSHDOWN_PARAMETERS if n in signature.parameters)


def effective_pushdown(
    declared: tuple[str, ...], request: SourceRequest
) -> tuple[str, ...]:
    """Narrow ``declared`` to the fields it is safe to hand over for this request.

    ``limit`` only composes if it is applied last. A source that truncates to
    ``limit`` rows and then has a filter applied -- or has its output sorted --
    returns the wrong rows, so whenever narrowing or ordering is still being
    enforced locally the limit comes back to the fallback with them.
    """
    narrowing_is_local = (request.filters and "filters" not in declared) or (
        request.order_by and "order_by" not in declared
    )
    if narrowing_is_local:
        return tuple(name for name in declared if name != "limit")
    return declared


def _request_parameter(signature: inspect.Signature) -> str | None:
    for name, parameter in signature.parameters.items():
        if name == "request" or parameter.annotation in (
            SourceRequest,
            "SourceRequest",
            "SourceRequest | None",
        ):
            return name
    return None


def bind(
    fn: Callable, request: SourceRequest, pushdown: tuple[str, ...] | None = None
) -> dict[str, Any]:
    """Keyword arguments to call ``fn`` with for this request.

    ``pushdown`` is the set of fields the function is allowed to own; anything
    outside it is withheld so the fallback stays authoritative. A field the
    caller left unset is omitted too, so the function's own default applies
    rather than being overwritten with None.
    """
    try:
        signature = inspect.signature(fn)
    except (TypeError, ValueError):
        return {}
    allowed = PUSHDOWN_PARAMETERS if pushdown is None else pushdown
    whole = _request_parameter(signature)
    if whole is not None:
        return {whole: _withhold(request, allowed)}
    unset = SourceRequest()
    return {
        name: getattr(request, name)
        for name in allowed
        if name in signature.parameters
        and getattr(request, name) != getattr(unset, name)
    }


def _withhold(request: SourceRequest, allowed: tuple[str, ...]) -> SourceRequest:
    """Blank the fields a whole-request function is not being trusted with."""
    if set(allowed) == set(PUSHDOWN_PARAMETERS):
        return request
    unset = SourceRequest()
    return replace(
        request,
        **{n: getattr(unset, n) for n in PUSHDOWN_PARAMETERS if n not in allowed},
    )


# --- fallback application ----------------------------------------------------


def apply(
    reader: pa.RecordBatchReader, request: SourceRequest, pushed_down: tuple[str, ...]
) -> pa.RecordBatchReader:
    """Enforce every contract field the function did not take ownership of.

    ``since`` and ``partition`` are semantic -- only the source knows what
    column carries its watermark or how it is partitioned -- so they have no
    fallback and are simply ignored when not pushed down.
    """
    schema = reader.schema
    filters = () if "filters" in pushed_down else request.filters
    columns = None if "columns" in pushed_down else request.columns
    limit = None if "limit" in pushed_down else request.limit
    order_by = () if "order_by" in pushed_down else request.order_by

    if filters:
        _validate_columns(schema, [f.column for f in filters], "filter on")
    if order_by:
        _validate_columns(schema, [s.column for s in order_by], "sort by")
    if columns:
        _validate_columns(schema, columns, "project")
        schema = pa.schema(
            [schema.field(name) for name in columns], metadata=schema.metadata
        )

    if not (filters or columns or order_by or limit is not None):
        return reader
    if order_by:
        # Sorting cannot stream: the last batch can hold the first row. This is
        # why a source that can order itself should claim ``order_by``.
        return _sorted_reader(reader, filters, order_by, columns, limit, schema)
    return pa.RecordBatchReader.from_batches(
        schema, _transform(reader, filters, columns, limit)
    )


def _sorted_reader(
    reader: pa.RecordBatchReader,
    filters: tuple[Filter, ...],
    order_by: tuple[Sort, ...],
    columns: tuple[str, ...] | None,
    limit: int | None,
    schema: pa.Schema,
) -> pa.RecordBatchReader:
    """Filter, then sort, then limit, then project -- in that order."""
    batches = [b for b in _transform(reader, filters, None, None)]
    table = (
        pa.Table.from_batches(batches, reader.schema)
        if batches
        else reader.schema.empty_table()
    )
    table = table.sort_by(
        [(s.column, "descending" if s.descending else "ascending") for s in order_by]
    )
    if limit is not None:
        table = table.slice(0, limit)
    if columns:
        table = table.select(list(columns))
    return pa.RecordBatchReader.from_batches(schema, table.to_batches())


def _validate_columns(
    schema: pa.Schema, columns: list[str] | tuple[str, ...], verb: str
) -> None:
    missing = [name for name in columns if name not in schema.names]
    if missing:
        raise ContractError(
            f"Cannot {verb} {', '.join(missing)}: not produced by this source. "
            f"Available columns: {', '.join(schema.names)}."
        )


def _transform(
    reader: pa.RecordBatchReader,
    filters: tuple[Filter, ...],
    columns: tuple[str, ...] | None,
    limit: int | None,
) -> Iterator[pa.RecordBatch]:
    emitted = 0
    for batch in reader:
        for predicate in filters:
            batch = batch.filter(_mask(batch, predicate))
        if columns:
            batch = batch.select(list(columns))
        if limit is not None:
            if emitted + batch.num_rows > limit:
                batch = batch.slice(0, limit - emitted)
            emitted += batch.num_rows
        if batch.num_rows:
            yield batch
        # Stop pulling once satisfied rather than draining the source.
        if limit is not None and emitted >= limit:
            return


def _mask(batch: pa.RecordBatch, predicate: Filter) -> pa.Array:
    import pyarrow.compute as pc

    column = batch.column(batch.schema.get_field_index(predicate.column))
    value = predicate.value
    if predicate.op == "=":
        return pc.equal(column, value)
    if predicate.op == "!=":
        return pc.not_equal(column, value)
    if predicate.op == ">":
        return pc.greater(column, value)
    if predicate.op == ">=":
        return pc.greater_equal(column, value)
    if predicate.op == "<":
        return pc.less(column, value)
    if predicate.op == "<=":
        return pc.less_equal(column, value)
    if predicate.op == "in":
        return pc.is_in(column, value_set=pa.array(_as_set(value)))
    if predicate.op == "not in":
        return pc.invert(pc.is_in(column, value_set=pa.array(_as_set(value))))
    if predicate.op == "like":
        return pc.match_like(column, str(value))
    raise ContractError(f"Unsupported filter operator {predicate.op!r}")


def _as_set(value: Any) -> list[Any]:
    if isinstance(value, (list, tuple)):
        return list(value)
    return [value]

File: crates/trilogy-io/src/contract.rs
//! What Trilogy may ask a source for, and how that ask is honored.
//!
//! Rust has no signature introspection, so a source declares what it pushes
//! down with [`Source::pushdown`](crate::Source::pushdown). Everything it does
//! not claim is enforced on the output stream by [`apply`], exactly as the
//! python wrapper does -- the contract is always honored either way.

use std::collections::BTreeMap;
use std::fmt;

use serde_json::Value;

use crate::error::{Error, Result};

pub const CONTRACT_VERSION: u32 = 1;

/// A contract field a source can take ownership of.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum Field {
    Limit,
    Columns,
    Filters,
    OrderBy,
    Since,
    Partition,
}

impl Field {
    pub fn as_str(&self) -> &'static str {
        match self {
            Field::Limit => "limit",
            Field::Columns => "columns",
            Field::Filters => "filters",
            Field::OrderBy => "order_by",
            Field::Since => "since",
            Field::Partition => "partition",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Op {
    Eq,
    NotEq,
    Lt,
    LtEq,
    Gt,
    GtEq,
    In,
    NotIn,
    Like,
}

impl Op {
    fn parse(raw: &str) -> Option<Op> {
        let normalized = raw.split_whitespace().collect::<Vec<_>>().join(" ");
        Some(match normalized.to_ascii_lowercase().as_str() {
            "=" => Op::Eq,
            "!=" => Op::NotEq,
            "<" => Op::Lt,
            "<=" => Op::LtEq,
            ">" => Op::Gt,
            ">=" => Op::GtEq,
            "in" => Op::In,
            "not in" => Op::NotIn,
            "like" => Op::Like,
            _ => return None,
        })
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Op::Eq => "=",
            Op::NotEq => "!=",
            Op::Lt => "<",
            Op::LtEq => "<=",
            Op::Gt => ">",
            Op::GtEq => ">=",
            Op::In => "in",
            Op::NotIn => "not in",
            Op::Like => "like",
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Filter {
    pub column: String,
    pub op: Op,
    pub value: Value,
}

impl fmt::Display for Filter {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {} {}", self.column, self.op.as_str(), self.value)
    }
}

impl Filter {
    /// Parses `<column> <op> <value>`, the same wire form the python side reads.
    ///
    /// The value is JSON first -- so numbers, lists and null work -- and a bare
    /// word falls back to a string.
    pub fn parse(text: &str) -> Result<Filter> {
        let text = text.trim();
        let (column, rest) = split_identifier(text)
            .ok_or_else(|| Error::contract(format!("Could not parse filter {text:?}")))?;
        let (op_text, value_text) = split_operator(rest)
            .ok_or_else(|| Error::contract(format!("Could not parse filter {text:?}")))?;
        let op = Op::parse(op_text)
            .ok_or_else(|| Error::contract(format!("Unsupported filter operator {op_text:?}")))?;
        Ok(Filter {
            column: column.to_string(),
            op,
            value: parse_value(value_text.trim()),
        })
    }
}

fn split_identifier(text: &str) -> Option<(&str, &str)> {
    let end = text
        .find(|c: char| !(c.is_alphanumeric() || c == '_' || c == '.'))
        .unwrap_or(text.len());
    if end == 0 {
        return None;
    }
    let (column, rest) = text.split_at(end);
    if !column.starts_with(|c: char| c.is_alphabetic() || c == '_') {
        return None;
    }
    Some((column, rest))
}

fn split_operator(text: &str) -> Option<(&str, &str)> {
    let text = text.trim_start();
    let lowered = text.to_ascii_lowercase();
    // Longest first, so `<=` never matches as `<` and `not in` never as `in`.
    for candidate in ["not in", ">=", "<=", "!=", "like", "in", "=", ">", "<"] {
        if lowered.starts_with(candidate) {
            let (op, rest) = text.split_at(candidate.len());
            // A word operator must be followed by a break, or `internal = 1`
            // would parse its column as `int` with operator `in`.
            if candidate.chars().next().is_some_and(|c| c.is_alphabetic())
                && !rest.starts_with(|c: char| c.is_whitespace() || c == '[' || c == '(')
            {
                continue;
            }
            return Some((op, rest));
        }
    }
    None
}

fn parse_value(raw: &str) -> Value {
    serde_json::from_str(raw)
        .unwrap_or_else(|_| Value::String(raw.trim_matches(['\'', '"']).to_string()))
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Sort {
    pub column: String,
    pub descending: bool,
}

impl Sort {
    /// Parses `<column>` or `<column>:asc` / `<column>:desc`.
    pub fn parse(text: &str) -> Result<Sort> {
        let (column, direction) = match text.split_once(':') {
            Some((c, d)) => (c, d.trim().to_ascii_lowercase()),
            None => (text, "asc".to_string()),
        };
        let descending = match direction.as_str() {
            "asc" => false,
            "desc" => true,
            _ => {
                return Err(Error::contract(format!(
                    "Could not parse sort {text:?}. Expected '<column>' or '<column>:asc' / '<column>:desc'."
                )))
            }
        };
        Ok(Sort {
            column: column.trim().to_string(),
            descending,
        })
    }

    pub fn render(&self) -> String {
        format!(
            "{}:{}",
            self.column,
            if self.descending { "desc" } else { "asc" }
        )
    }
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct SourceRequest {
    pub limit: Option<usize>,
    pub columns: Option<Vec<String>>,
    pub filters: Vec<Filter>,
    pub order_by: Vec<Sort>,
    pub since: Option<String>,
    pub partition: BTreeMap<String, String>,
}

impl SourceRequest {
    fn is_set(&self, field: Field) -> bool {
        match field {
            Field::Limit => self.limit.is_some(),
            Field::Columns => self.columns.is_some(),
            Field::Filters => !self.filters.is_empty(),
            Field::OrderBy => !self.order_by.is_empty(),
            Field::Since => self.since.is_some(),
            Field::Partition => !self.partition.is_empty(),
        }
    }

    fn clear(&mut self, field: Field) {
        match field {
            Field::Limit => self.limit = None,
            Field::Columns => self.columns = None,
            Field::Filters => self.filters.clear(),
            Field::OrderBy => self.order_by.clear(),
            Field::Since => self.since = None,
            Field::Partition => self.partition.clear(),
        }
    }

    /// The request as handed to a source: fields it does not own are blanked,
    /// so the fallback stays authoritative.
    pub fn withheld(&self, pushdown: &[Field]) -> SourceRequest {
        let mut request = self.clone();
        for field in [
            Field::Limit,
            Field::Columns,
            Field::Filters,
            Field::OrderBy,
            Field::Since,
            Field::Partition,
        ] {
            if !pushdown.contains(&field) {
                request.clear(field);
            }
        }
        request
    }
}

/// Narrow a declared pushdown to what is safe to hand over for this request.
///
/// `limit` only composes if it is applied last. A source that truncates to
/// `limit` rows and then has a filter applied -- or has its output sorted --
/// returns the wrong rows, so whenever narrowing or ordering is still enforced
/// locally the limit comes back to the fallback with it.
pub fn effective_pushdown(declared: &[Field], request: &SourceRequest) -> Vec<Field> {
    let local = |field: Field| request.is_set(field) && !declared.contains(&field);
    let narrowing_is_local = local(Field::Filters) || local(Field::OrderBy);
    declared
        .iter()
        .copied()
        .filter(|field| !(narrowing_is_local && *field == Field::Limit))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_operators_and_values() {
        assert_eq!(
            Filter::parse("i >= 5").unwrap(),
            Filter {
                column: "i".into(),
                op: Op::GtEq,
                value: Value::from(5)
            }
        );
        assert_eq!(
            Filter::parse("state = CA").unwrap().value,
            Value::from("CA")
        );
        assert_eq!(
            Filter::parse(r#"state = "CA""#).unwrap().value,
            Value::from("CA")
        );
        assert_eq!(Filter::parse("s NOT IN [1]").unwrap().op, Op::NotIn);
        assert_eq!(Filter::parse("s like n%").unwrap().value, Value::from("n%"));
        assert_eq!(
            Filter::parse(r#"state in ["CA","NY"]"#).unwrap().value,
            serde_json::json!(["CA", "NY"])
        );
    }

    #[test]
    fn a_column_starting_with_an_operator_word_still_parses() {
        let filter = Filter::parse("internal = 1").unwrap();
        assert_eq!(filter.column, "internal");
        assert_eq!(filter.op, Op::Eq);
    }

    #[test]
    fn rejects_garbage() {
        assert!(Filter::parse("this is not a filter").is_err());
        assert!(Filter::parse("").is_err());
    }

    #[test]
    fn limit_pushdown_is_withdrawn_when_filters_stay_local() {
        let request = SourceRequest {
            limit: Some(4),
            filters: vec![Filter::parse("state = CA").unwrap()],
            ..Default::default()
        };
        assert_eq!(effective_pushdown(&[Field::Limit], &request), vec![]);
        assert_eq!(
            effective_pushdown(&[Field::Limit, Field::Filters], &request),
            vec![Field::Limit, Field::Filters]
        );
        let plain = SourceRequest {
            limit: Some(4),
            ..Default::default()
        };
        assert_eq!(
            effective_pushdown(&[Field::Limit], &plain),
            vec![Field::Limit]
        );
    }

    #[test]
    fn parses_sorts() {
        assert_eq!(
            Sort::parse("id:desc").unwrap(),
            Sort {
                column: "id".into(),
                descending: true
            }
        );
        assert_eq!(
            Sort::parse("id").unwrap(),
            Sort {
                column: "id".into(),
                descending: false
            }
        );
        assert!(Sort::parse("id:sideways").is_err());
        assert_eq!(Sort::parse("id:desc").unwrap().render(), "id:desc");
    }

    #[test]
    fn a_local_ordering_also_keeps_the_limit_home() {
        let request = SourceRequest {
            limit: Some(4),
            order_by: vec![Sort {
                column: "id".into(),
                descending: true,
            }],
            ..Default::default()
        };
        assert_eq!(effective_pushdown(&[Field::Limit], &request), vec![]);
        assert_eq!(
            effective_pushdown(&[Field::Limit, Field::OrderBy], &request),
            vec![Field::Limit, Field::OrderBy]
        );
    }

    #[test]
    fn withheld_blanks_unowned_fields() {
        let request = SourceRequest {
            limit: Some(4),
            filters: vec![Filter::parse("state = CA").unwrap()],
            ..Default::default()
        };
        let handed = request.withheld(&[Field::Filters]);
        assert!(handed.limit.is_none());
        assert_eq!(handed.filters.len(), 1);
    }
}
